Include default password in CSV [LEI-253]

samanehsan opened the pull request:
Refs: https://openscience.atlassian.net/browse/LEI-253
## Purpose

When creating accounts without an additional field used as the password, include the automatically generated password in the CSV file of participant accounts.
## Summary of changes

Add `password` as a field in `extra` and set `useAsPassword` to `password`
## Testing notes

To test, create an account without adding any additional fields and check that there is a "password" column in the resulting CSV with account information

samanehsan committed:
Include default password in CSV

When creating accounts without an
additional field used as the password,
include the automatically generated
password in the CSV file of participant
accounts.

abought commented on app/components/participant-creator/component.js:
```diff
@@ -73,6 +73,12 @@ export default Ember.Component.extend({
+                    value: password
+                });
+                extra['password'] = password;
+                this.set('useAsPassword', 'password');
```
This parameter is used like a boolean in the template, but the value is just the string "password".

(it's neither a boolean, nor an actual password)

Is this intentional?

samanehsan replied:
It looks to be intentional -- `useAsPassword` is the name of the field that should be used as the password

abought replied:
Does that imply that our accounts have a default password of "password"?

We should not do that. Preliminary experiments show that accounts actually get a randomly generated password, which is much better- but it means that this value is confusing.

samanehsan replied:
`useAsPassword` is the name of the field whose value should be used as the password, not the actual password. Maybe `passwordField` is more clear?

abought replied:
So instead of storing the password in one place, we use a field.... to track the name of another configurable field... that stores the password?

Is there perhaps a more direct approach?

samanehsan replied:
@abought, is it okay if I refactor how the password gets stored in a follow-up PR? That way, it does not need to block the rest of the changes in this PR.

samanehsan replied:
Here is the follow-up ticket for refactoring `useAsPassword`: https://openscience.atlassian.net/browse/LEI-280

abought commented on app/components/participant-creator/template.hbs:
```diff
@@ -45,7 +45,7 @@
                                         Use as password?
                                     </label>
-                                    <input name="useAsPassword" onchange={{action (action 'useAsPassword' item.key) value="target.checked"}} type="radio" />
+                                    <input name="useAsPassword" checked={{if useAsPassword true}} onchange={{action (action 'useAsPassword' item.key) value="target.checked"}} type="radio" />
```
It is slightly confusing to have both a variable and an action by the same name in this controller. Can you please rename one to help differentiate?

abought replied:
Can you please clarify why this is a radio button instead of a checkbox?

abought replied:
I tried to create one account, tag `ATAG`, Additional fields `A-field`

I then told it to use that extra field as password, and clicked "create accounts".

I got the following error on the console, then the page hung- it does not seem to deal with failures on the server.

```
Saving nOte4-ATAG
bcrypt.js:172 Uncaught Error: Illegal arguments: object, string
```

abought reviewed (changes requested):
I am having trouble running this feature when I manually test locally. Can you please take a look before this moves forward?

samanehsan committed:
Use checkbox instead of radio button

samanehsan committed:
Rename useAsPassword action

Rename useAsPassword action to
differentiate it from the variable
with the same name.

samanehsan committed:
Handle invalid password

Handle error where user specifies a field
to use as the password but does not enter
a value.

abought commented:
Two possible bugs from initial functional review:

**Edge case where accounts are created without a password saved**

I created 3 accounts with default password. It auto-generated a password, and added that to "additional fields" on the page. (presumably the same password for all 3 accounts?)

Then, I cleared the additional field using the yellow x, and clicked create again. The next CSV file to be saved did not have a password column at all.

**In the first CSV file, all 3 newly created accounts listed the same exact password. That seems to be not very secure.**

I have not yet tried to actually log in to all 3 newly created accounts with the information in the CSV file, but I encourage you to do so to verify that the account info is being saved correctly.

samanehsan committed:
Fix issue creating default passwords

Clear useAsPassword and password extra
field each time the default password is
used when creating accounts. Fixes issue
when creating two batches of accounts that
use default passwords and only the first
batch passords get written to the csv.

samanehsan commented:
The UI changes in https://github.com/CenterForOpenScience/experimenter/pull/119 make "Study ID" a separate, required field so it will not be possible to create accounts that do not have a password.
